nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java

```diff
@@ -90,6 +93,15 @@ public void onEnabled(final ConfigurationContext context) throws InitializationE
         try {
             setupClient(context);
             responseCharset = Charset.forName(context.getProperty(CHARSET).getValue());
+
+            // re-create the ObjectMapper in case the SUPPRESS_NULLS property has changed - the JsonInclude settings aren't dynamic
+            mapper = new ObjectMapper();
+            if (ALWAYS_SUPPRESS.getValue().equals(context.getProperty(SUPPRESS_NULLS).getValue())) {
+                mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
```

Comment:
I ran this with Never once and Always once, and both times with a field `y` set to `null` it doesn't appear to get into the document. I expect that for Always but if I query the document that had Never Suppress set and one field (that is non-null in other docs in the index) that is null, shouldn't I expect `"y": null` in the document returned by the query?

Reply:
You are correct, I would expect the `"y": null` output when using `NEVER_SUPPRESS` (but not for `ALWAYS_SUPPRESS`).

I've included this as a test in the `ElasticSearch5ClientService_IT` as `testNullSuppression`, which involves:
* creating a document with `NEVER_SUPPRESS` that contains and asserts the inserted document matches:
  * "is_null": null
  * "is_empty": empty ("")
  * "is_blank": blank (" ")
  * "msg": "test"
* creating the same document with `ALWAYS_SUPPRESS` then asserting that the document *does not* contain the `"is_null"` or `"is_empty"` fields

Switching between the two states involves disabling the `ElasticSearchClientService`, changing the `SUPPRESS_NULLS` property, then re-enabling the service before using it to to `bulk` index the document then `get` it back from Elasticsearch.


How are you running the code and checking the output if not via the tests?

Reply:
I also temporarily logged the body being sent to Elasticsearch within `bulk` and the response body from Elasticsearch in `get` prior to the ObjectMapper being used... both showed the expected lack of `null` value fields with `ALWAYS` but inclusion of the same with `NEVER`.

Reply:
For the behavior I saw above, I compiled and built the ES NARs, dropped them into a NiFi based off `main`, and ran against a Docker container running ES 5.6.16. Using PutElasticsearchRecord I wasn't seeing the `"y": null` when set to Never Suppress. I forgot to look at the current (main not this PR) behavior to see if the "y" was in the document returned by the query, but it wasn't with this PR and the aforementioned setup.

Reply:
I tried the same thing with ES 7.9.3 and I get the same results. I even used an explicit schema saying `y` is a nullable field, but when I sent in a document with `"y": null` and then query it with `curl 'http://localhost:9200/nulltest/_search?q=*'`, I get no `y` element in that doc. Integration test passes (when using Maven) though...

Reply:
I agree, for some reason the behaviour of the controller service once built is different to that in the integration tests (which is kind of disturbing in several ways).

Will investigate the lack of `null` values when using the `NEVER_SUPPRESS` option. If possible, I'll also try to see whether there's a general issue with the integration tests.

Reply:
I couldn't re-produce the `null` scenario I'm sure I saw while trying to use the PutElasticsearchRecord processor in 1.11.4.

However, I do still think there's something to change here, in that you can't currently output `null`/empty values to Elasticsearch using this processor (so kind of the reverse of the original ticket) due to the `removeEmpty` method in the `PutElasticsearchRecord` processor. Looking at this method, it could also result in empty nested objects/arrays being sent to Elasticsearch currently, which wouldn't always be ideal.

So, I've changed the code to:
* Not `removeEmpty` within `PutElasticsearchRecord`
* Use the Jackson ObjectMapper to remove `null`/empty fields (including empty nested arrays/objects) when `ALWAYS_SUPPRESS` is selected
* Leave the `null`/empty fields in the output to Elasticsearch when `NEVER_SUPPRESS` is selected

I've added a Flow template/export to the Jira ticket (included converting the data into Avro as that's what I use in the Flow where I believe I saw this behaviour in 1.11.4 before switching to the `PutElasticsearchHttpRecord` instead - although that didn't cause a re-production of the issue either).

Re-based the branch from `main` and updated both unit and integration tests to cover these scenarios. Default behaviour of the `PutElasticsearchRecord` processor remains one of removing `null`/empty fields (i.e. `ALWAYS_SUPPRESS` in `ElasticSearchClientService`). I think my concerns with the integration tests above are explained by the fact that testing the service directly wasn't executing the `remoteEmpty` method of the processor.

Reply:
@mattyb149 managed to confirm that this is a problem for my flow in 1.11.4 when outputting data to Elasticsearch that is a Record containing nested Records with Arrays or Records in them... the fields within the nested Records come through as `null` when that's not wanted. If it makes a difference, we're using Avro as the data format between processors in this Flow.

I think the logic in PutElasticsearchRecord is missing an if-else to handle `Object[]` as well as `List` and/or other structures (the `DataTypeUtils` to convert between NiFi Record structure and "plain" Java Objects is a little flawed with some of these conversions - I've recently raised separate Jira tickets documenting some of those).

The switch to using the Jackson-based null/empty suppression should allow for better null field control.